site/_posts/2020-08-07-locking-an-account-with-breached-password.md

```diff
@@ -0,0 +1,258 @@
+  error_log("Invalid authorization header value found: ".$authorization_value);
+  return;
+}
+//...
+```
+
+Then, the code checks the `Authorization` header. This ensures that only FusionAuth calls this webhook. For production, you would definitely want to use TLS as well.
+
+```php
+//...
+$input = file_get_contents('php://input');
+
+$obj = json_decode($input);
+//...
+```
+
+In these lines, the entire contents of the webhook payload are converted into a string. The string is then decoded into a JSON object for easier handling.
+
+```php
+//...
+if (!$obj) { 
+  error_log("Invalid JSON");
+  return;
+}
+
+type = $obj->event->type;
+if ($type !== "user.password.breach") {
+  error_log("Sorry, we only handle breached password events.");
+  return;
+}
+
+$user_id = $obj->event->user->id;
+if (!$user_id) {
+  error_log("No user id");
+  return;
+}
+//...
+```
+
+Next, validate the payload. If you don't get valid JSON, a password breach event, and a user id, simply log an error and return. Doing so allows the webhook and the login event to succeed. Check for the event type just in case there's a misconfiguration and our webhook is notified of other types of events.
+
+```php
+//...
+$client = new FusionAuth\FusionAuthClient($api_key, $fa_url);
+$response = $client->deactivateUser($user_id);
+if (!$response->wasSuccessful()) {
+  error_log("Response wasn't successful:");
+  error_log(var_export($response, TRUE));
+  return;
+}
+//...
+```
+
+Finally! This is where the action is. This code creates a new FusionAuth client. It then deactivates the user who logged in with a password found to be compromised. You could take other steps here, too. 
+
+You could do more within FusionAuth, by, for example:
+
+* Adding a date of deactivation to the `user.data` field
+* [Actioning the user](/docs/v1/tech/apis/actioning-users) for display in the administrative interface or to be queried via the API
+* Putting the user in a [group](/docs/v1/tech/core-concepts/groups) for future processing
+
+You could also integrate with other systems.
+
+* You could fire off an API call to another service that needs to know about this security violation.
+* The system could add an event to a streaming service, such as Kafka, for future analysis.
+* Your application could send an email to the user and their boss about the situation. Wouldn't be cool, but you could do it.
+
+```php
+//...
+http_response_code(500);
+//...
+```
+
+Finally, the code returns a `500` HTTP status code. This stops the login process. Because you configured this tenant to require all webhooks to succeed before processing the event, if any don't, the event doesn't complete. This means the user is not logged in. 
+
+If this webhook didn't fail, the user would be logged in. The account would be deactivated; they'd be unable to login later. But their current session would be active for the duration of the token. We don't want that to happen, so that's why the code returns a `500`.
```

Comment:
Up to you, if you want to further clarify this. Any non `2xx` status code will be considered a failure. But `500` is just fine.

Reply:
Ah, the docs aren't consistent.

See https://fusionauth.io/docs/v1/tech/events-webhooks/writing-a-webhook : "If there was any type of error or failure, your Webhook must send back an HTTP response status code of 500" vs https://fusionauth.io/docs/v1/tech/events-webhooks/ : "The event will succeed as long as one or more of the webhooks respond with a status code between 200 and 299 (inclusive)."

I'll modify the former.

Reply:
Yeah, you're correct - I think we did document it like that - probably best to keep it then. In "practice" we are only looking for a non `2xx` - but just fine to keep it stated that way.